pkg/controller/remotecluster/controller.go

```diff
@@ -0,0 +1,250 @@
+
+	c := &Controller{
+		rcMgrCache:                NewCache(),
+		UUID:                      uuid,
+		kubeClient:                kubeClient,
+		ramaClient:                ramaClient,
+		remoteClusterLister:       remoteClusterInformer.Lister(),
+		remoteClusterSynced:       remoteClusterInformer.Informer().HasSynced,
+		remoteSubnetLister:        remoteSubnetInformer.Lister(),
+		remoteSubnetSynced:        remoteSubnetInformer.Informer().HasSynced,
+		localClusterSubnetLister:  localClusterSubnetInformer.Lister(),
+		localClusterSubnetSynced:  localClusterSubnetInformer.Informer().HasSynced,
+		remoteVtepLister:          remoteVtepInformer.Lister(),
+		remoteVtepSynced:          remoteSubnetInformer.Informer().HasSynced,
+		localClusterNetworkLister: localClusterNetworkInformer.Lister(),
+		localClusterNetworkSynced: localClusterNetworkInformer.Informer().HasSynced,
+		remoteClusterQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+		rcMgrQueue:                workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "remoteclustermanager"),
+		recorder:                  recorder,
+	}
+
+	remoteClusterInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: c.filterRemoteCluster,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addOrDelRemoteCluster,
+			UpdateFunc: c.updateRemoteCluster,
+			DeleteFunc: c.addOrDelRemoteCluster,
+		},
+	})
+
+	return c
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) error {
+	defer runtimeutil.HandleCrash()
+	defer c.rcMgrQueue.ShutDown()
+	defer c.remoteClusterQueue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.remoteClusterSynced, c.remoteSubnetSynced, c.remoteVtepSynced, c.localClusterSubnetSynced); !ok {
+		return fmt.Errorf("%s failed to wait for caches to sync", ControllerName)
+	}
+
+	// start workers
+	klog.Info("Starting workers")
+	go wait.Until(c.runRemoteClusterWorker, time.Second, stopCh)
+	go wait.Until(c.processRCManagerQueue, time.Second, stopCh)
+	go wait.Until(c.runOverlayNetIDWorker, time.Minute, stopCh)
+	go wait.Until(c.updateRemoteClusterStatus, HealthCheckPeriod, stopCh)
+	<-stopCh
+
+	c.closeRemoteClusterManager()
+
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *Controller) closeRemoteClusterManager() {
+	// no need to lock
+	for _, mgr := range c.rcMgrCache.rcMgrMap {
+		mgr.Close()
+	}
+}
+
+func (c *Controller) runOverlayNetIDWorker() {
+	c.overlayNetIDMU.Lock()
+	defer c.overlayNetIDMU.Unlock()
+
+	networks, err := c.localClusterNetworkLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("Can't list local cluster network. err=%v", err)
+		return
+	}
+	for _, network := range networks {
+		if network.Spec.Type == networkingv1.NetworkTypeOverlay {
+			n := network.DeepCopy()
+			c.OverlayNetID = n.Spec.NetID
```

Comment:
loop here is not effective enough than events

Reply:
I don't think there are too much difference for it uses the lister. By the way, events needs more code and is not as clear as single function.

Reply:
you mean `epoll` has no difference with `select`?

Reply:
how many networks can you have in one cluster?